src/pybamm/util.py

```diff
@@ -108,38 +109,98 @@ def __getitem__(self, key):
                 f"'{key}' not found. Best matches are {best_matches}"
             ) from error
 
-    def search(self, key, print_values=False):
+    def _find_matches(self, search_key: str, known_keys: list[str]):
         """
-        Search dictionary for keys containing 'key'. If print_values is True, then
-        both the keys and values will be printed. Otherwise just the values will
-        be printed. If no results are found, the best matches are printed.
+        Helper method to find exact and partial matches for a given search key.
+
+        Parameters
+        ----------
+        search_key : str
+            The term to search for in the keys.
+        known_keys : list of str
+            The list of known dictionary keys to search within.
+
         """
-        key_in = key
-        key = key_in.lower()
+        exact = [key for key in known_keys if search_key in key.lower()]
+        partial = difflib.get_close_matches(search_key, known_keys, n=5, cutoff=0.5)
+        return exact, partial
 
-        # Sort the keys so results are stored in alphabetical order
-        keys = list(self.keys())
-        keys.sort()
-        results = {}
+    def search(self, keys: str | list[str], print_values: bool = False):
+        """
+        Search dictionary for keys containing all terms in 'keys'.
+        If print_values is True, both the keys and values will be printed.
+        Otherwise, just the keys will be printed. If no results are found,
+        the best matches are printed.
+
+        Parameters
+        ----------
+        keys : str or list of str
+            Search term(s)
+        print_values : bool, optional
+            If True, print both keys and values. Otherwise, print only keys.
+            Default is False.
+        """
 
-        # Check if any of the dict keys contain the key we are searching for
-        for k in keys:
-            if key in k.lower():
-                results[k] = self[k]
+        if not isinstance(keys, (str, list)) or not all(
+            isinstance(k, str) for k in keys
+        ):
+            msg = f"'keys' must be a string or a list of strings, got {type(keys)}"
+            raise TypeError(msg)
 
-        if results == {}:
-            # If no results, return best matches
-            best_matches = self.get_best_matches(key)
+        if isinstance(keys, str):
+            if not keys.strip():
+                raise ValueError(
+                    "The search term cannot be an empty or whitespace-only string"
+                )
+            original_keys = [keys]
+            search_keys = [keys.strip().lower()]
+
+        elif isinstance(keys, list):
+            if all(not str(k).strip() for k in keys):
+                raise ValueError(
+                    "The 'keys' list cannot contain only empty or whitespace strings"
+                )
```

Comment:
```suggestion
                msg = "The 'keys' list cannot contain only empty or whitespace strings"
                raise ValueError(msg)
```

Reply:
I considered making this change earlier but assumed that this format of writing error messages was preferred for f-strings. Thanks for pointing it out! I'll implement this now to maintain consistency in the code.